include/StochasticNoiseSimulator.hpp

```diff
@@ -3,189 +3,142 @@
 
 #include "QuantumComputation.hpp"
 #include "Simulator.hpp"
+#include "dd/NoiseFunctionality.hpp"
 
-#include <cstddef>
-#include <map>
-#include <memory>
-#include <random>
-#include <string>
+#include <optional>
 #include <thread>
 #include <vector>
 
-class StochasticNoiseSimulator: public Simulator<StochasticNoiseSimulatorDDPackage> {
+template<class DDPackage = StochasticNoisePackage>
+class StochasticNoiseSimulator: public Simulator<DDPackage> {
 public:
-    StochasticNoiseSimulator(std::unique_ptr<qc::QuantumComputation>& qc, const unsigned int step_number, const double step_fidelity):
-        qc(qc), step_number(step_number), step_fidelity(step_fidelity) {
-        dd->resize(qc->getNqubits());
+    StochasticNoiseSimulator(std::unique_ptr<qc::QuantumComputation>& qc, const unsigned int stepNumber, const double stepFidelity):
+        qc(qc), stepNumber(stepNumber), stepFidelity(stepFidelity) {
+        Simulator<DDPackage>::dd->resize(qc->getNqubits());
+        setMaxInstances();
     }
 
-    StochasticNoiseSimulator(std::unique_ptr<qc::QuantumComputation>& qc, const unsigned int step_number, const double step_fidelity, unsigned long long seed):
-        Simulator(seed), qc(qc), step_number(step_number), step_fidelity(step_fidelity) {
-        dd->resize(qc->getNqubits());
+    StochasticNoiseSimulator(std::unique_ptr<qc::QuantumComputation>& qc, const unsigned int stepNumber, const double stepFidelity, std::size_t seed):
+        Simulator<DDPackage>(seed), qc(qc), stepNumber(stepNumber), stepFidelity(stepFidelity) {
+        Simulator<DDPackage>::dd->resize(qc->getNqubits());
+        setMaxInstances();
     }
 
     StochasticNoiseSimulator(std::unique_ptr<qc::QuantumComputation>& qc,
-                             const std::string&                       noise_effects,
-                             double                                   noise_prob,
-                             long                                     stoch_runs,
-                             unsigned int                             step_number,
-                             double                                   step_fidelity,
-                             const std::string&                       recorded_properties):
-        qc(qc),
-        step_number(step_number), step_fidelity(step_fidelity) {
-        setNoiseEffects(noise_effects);
-        setRecordedProperties(recorded_properties);
-        initializeNoiseProbabilities(noise_prob);
-        setNumberOfRuns(stoch_runs);
-    }
+                             const std::string&                       cNoiseEffects,
+                             double                                   cNoiseProbability,
+                             std::optional<double>                    cAmpDampingProbability,
+                             double                                   cMultiQubitGateFactor,
+                             std::size_t                              cStochRuns,
```

Comment:
no need for the `c` in front of the parameters

Reply:
Changed that